File: java/dataformatter/src/test/java/com/cefalo/oopcourse/experiment/CSVFormatterTest.java
package com.cefalo.oopcourse.experiment;


import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;

/**
 * Test case for testing the csv formatter
 */
public class CSVFormatterTest {

    @Mock
    DataService dataServiceMock;

    @Rule
    public MockitoRule mockitoRule = MockitoJUnit.rule();

    @Test
    public void testCsv() throws Exception {
        CSVFormatter t = new CSVFormatter(dataServiceMock);
        when(dataServiceMock.readData()).thenReturn(DataService.createDummyItems());

        t.execute();

        verify(dataServiceMock, times(1)).openConnection();

        String expectedText = getExpectedFormat();
        verify(dataServiceMock).exportData(eq(expectedText));

        verify(dataServiceMock, times(1)).closeConnection();
    }

    private String getExpectedFormat() {
        return "1, First Test item, 100.0\n" +
                "2, second Test item, 210.0\n" +
                "3, third Test item, 200.0\n" +
                "4, fourth Test item, 300.0\n";
    }
}

File: dotnet/DataFormatter/DataFormetter.Tests/CSVFormatterTest.cs
using System;
using DataFormatter;
using DataFormatter.Common;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;

namespace DataFormetter.Tests
{
    [TestClass]
    public class CSVFormatterTest
    {
        private readonly Mock<IDataService> _dataServiceMock = new Mock<IDataService>();
        [TestMethod]
        public void TestCsv()
        {
            CSVFormatter t = new CSVFormatter(_dataServiceMock.Object);
            var dummyItems = DataService.CreateDummyItems();
            _dataServiceMock.Setup(d => d.ReadData()).Returns(dummyItems);

            t.Execute();
            _dataServiceMock.Verify(d => d.OpenConnection(), Times.Once);

            string expectedText = getExpectedFormat();
            _dataServiceMock.Verify(d => d.ExportData(expectedText), Times.Once);
            _dataServiceMock.Verify(d => d.CloseConnection(), Times.Once);
        }

        private String getExpectedFormat()
        {
            return "1, First Test item, 100\n" +
            "2, second Test item, 210\n" +
            "3, third Test item, 200\n" +
            "4, fourth Test item, 300\n";
        }
    }
}
